Accept file `URL` instances

jopemachine opened the pull request:
Fixes #36.

sindresorhus commented on index.js:
```diff
@@ -43,6 +44,10 @@ const permissionError = pth => {
 };
 
+const getTargetDir = input => {
+	return typeof input === 'string' ? path.resolve(input) : input;
```
What if the URL is relative? This is not the correct way to handle `URL`.

sindresorhus commented on index.js:
```diff
@@ -51,7 +56,7 @@ const makeDir = async (input, options) => {
 
 	if (useNativeRecursiveOption && options.fs.mkdir === fs.mkdir) {
-		const pth = path.resolve(input);
+		const pth = getTargetDir(input);
```
The `.resolve` call is moot anyway here I think as the native method resolves internally.

sindresorhus commented:
You're not handling `URL` where it's actually needed, on line 106.

jopemachine commented:
Removed moot function call and added some tests for relative url.
It seems working for relative urls.
I would appreciate your letting me know if I got wrong :)

sindresorhus commented:
> You're not handling URL where it's actually needed, on line 106.

^

jopemachine commented:
Done. I'm sorry for the missing.

sindresorhus commented on index.js:
```diff
@@ -32,6 +36,10 @@ const processOptions = options => {
 };
 
+const getTargetDir = input => {
+	return path.resolve(typeof input === 'string' ? input : input.pathname);
```
This is incorrect. The Node.js docs even point that out: https://nodejs.org/api/url.html#urlfileurltopathurl

jopemachine replied:
> This is incorrect. The Node.js docs even point that out: [nodejs.org/api/url.html#urlfileurltopathurl](https://nodejs.org/api/url.html#urlfileurltopathurl)

Thanks for letting me know. It seems I need more study Node modules.

Could I ask if we could use `fileURLToPath` here?
The document says it was added in 10.12.0, but this module's target node engines is greater than or equal to 10.
Is it acceptable to change this module's engine to 10.12.0?
Or do I need to add some polyfills to replace `fileURLToPath`?

sindresorhus replied:
It's ok to use `fileURLToPath`. I plan to target Node.js 14 soon.

sindresorhus commented on index.js:
```diff
@@ -98,7 +104,7 @@ const makeDir = async (input, options) => {
 	};
 
-	return make(path.resolve(input));
+	return make(getTargetDir(input));
```
Don't use abbreviations.

sindresorhus commented on index.js:
```diff
@@ -32,6 +37,10 @@ const processOptions = options => {
 	};
 };
 
```
Use a helper method for readability:

```js
const toPath = urlOrPath => urlOrPath instanceof URL ? fileURLToPath(urlOrPath) : urlOrPath;
```

sindresorhus commented:
Tests were failing on Windows.
